### PR TITLE
DROOLS-3199: [DMN Designer] Clear properties panel when multiple cells are selected

### PR DESCRIPTION
**Pull request by manstis**
See https://issues.jboss.org/browse/DROOLS-3199

### CODE REVIEW

**Comment by jomarko** [comment]
@manstis no comments to code, finished first round of review as discussed on slack.

**Comment by manstis** [comment]
@jomarko Updates made.

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/editors/expressions/types/context/ContextGridTest.java [hunk 1]
```diff
@@ -858,6 +858,40 @@ public void testSelectRow() {
...
+    @SuppressWarnings("unchecked")
+    public void testSelectMultipleRows() {
+        setupGrid(0);
+
+        grid.selectCell(0, ContextUIModelMapperHelper.ROW_COLUMN_INDEX, false, false);
+
+        assertDomainObjectSelection(expression.get().getContextEntry().get(0).getVariable());
+
+        //Reset DomainObjectSelectionEvent tested above.
+        reset(domainObjectSelectionEvent);
```
Comment (karreiro) [inline kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/editors/expressions/types/context/ContextGridTest.java]: I think that the need of resetting the `domainObjectSelectionEvent` is a smell that we could split this test in two methods.. like: `testSelectMultipleCellsWhenControlIsKeyDown` and `testSelectMultipleCellsWhenControlIsNotKeyDown`.

..but this is a super-minor, feel free to keep it as is.

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/editors/expressions/types/context/ContextGridTest.java [hunk 2]
```diff
@@ -858,6 +858,40 @@ public void testSelectRow() {
...
+    @SuppressWarnings("unchecked")
+    public void testSelectSingleRowWithHeaderSelected() {
+        setupGrid(0);
+
+        grid.selectHeaderCell(0, ContextUIModelMapperHelper.NAME_COLUMN_INDEX, false, false);
+
+        assertDomainObjectSelection(hasExpression);
+
+        //Reset DomainObjectSelectionEvent tested above.
+        reset(domainObjectSelectionEvent);
```
Comment (karreiro) [inline kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/editors/expressions/types/context/ContextGridTest.java]: ..the same comment above.

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/editors/expressions/types/dtable/DecisionTableGridTest.java [hunk 3]
```diff
@@ -1220,6 +1220,40 @@ public void testSelectRow() {
...
+    @SuppressWarnings("unchecked")
+    public void testSelectMultipleCells() {
+        setupGrid(makeHasNameForDecision(), 0);
+
+        grid.selectCell(0, DEFAULT_INPUT_CLAUSE_COLUMN_INDEX, false, false);
+
+        assertDomainObjectSelection(expression.get().getRule().get(0).getInputEntry().get(0));
+
+        //Reset DomainObjectSelectionEvent tested above.
+        reset(domainObjectSelectionEvent);
```
Comment (karreiro) [inline kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/editors/expressions/types/dtable/DecisionTableGridTest.java]: ..the same comment above.

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/editors/expressions/types/dtable/DecisionTableGridTest.java [hunk 4]
```diff
@@ -1220,6 +1220,40 @@ public void testSelectRow() {
...
+    @SuppressWarnings("unchecked")
+    public void testSelectSingleCellWithHeaderSelected() {
+        setupGrid(makeHasNameForDecision(), 0);
+
+        grid.selectHeaderCell(0, DEFAULT_INPUT_CLAUSE_COLUMN_INDEX, false, false);
+
+        assertDomainObjectSelection(expression.get().getInput().get(0));
+
+        //Reset DomainObjectSelectionEvent tested above.
+        reset(domainObjectSelectionEvent);
```
Comment (karreiro) [inline kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/editors/expressions/types/dtable/DecisionTableGridTest.java]: ..the same comment above.

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/editors/expressions/types/relation/RelationGridTest.java [hunk 5]
```diff
@@ -832,6 +832,57 @@ public void testSelectRow() {
...
+
+        grid.selectCell(0, 1, false, false);
+
+        verify(domainObjectSelectionEvent).fire(domainObjectSelectionEventCaptor.capture());
+
+        final DomainObjectSelectionEvent event1 = domainObjectSelectionEventCaptor.getValue();
+        assertThat(event1.getDomainObject()).isEqualTo(expression.get().getRow().get(0).getExpression().get(0));
+
+        //Reset DomainObjectSelectionEvent tested above.
+        reset(domainObjectSelectionEvent);
```
Comment (karreiro) [inline kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/editors/expressions/types/relation/RelationGridTest.java]: ..the same comment above.

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/editors/expressions/types/relation/RelationGridTest.java [hunk 6]
```diff
@@ -832,6 +832,57 @@ public void testSelectRow() {
...
+
+        grid.selectHeaderCell(0, RelationUIModelMapperHelper.ROW_INDEX_COLUMN_COUNT, false, false);
+
+        verify(domainObjectSelectionEvent).fire(domainObjectSelectionEventCaptor.capture());
+
+        final DomainObjectSelectionEvent event1 = domainObjectSelectionEventCaptor.getValue();
+        assertThat(event1.getDomainObject()).isEqualTo(expression.get().getColumn().get(0));
+
+        //Reset DomainObjectSelectionEvent tested above.
+        reset(domainObjectSelectionEvent);
```
Comment (karreiro) [inline kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/editors/expressions/types/relation/RelationGridTest.java]: ..the same comment above.

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/expressions/types/context/ContextGrid.java [hunk 7]
```diff
@@ -321,6 +322,15 @@ void clearExpressionType(final int uiRowIndex) {
...
     protected void doAfterSelectionChange(final int uiRowIndex,
                                           final int uiColumnIndex) {
+        if (getModel().getSelectedCells().stream().map(GridData.SelectedCell::getRowIndex).distinct().count() > 1) {
+            super.doAfterSelectionChange(uiRowIndex, uiColumnIndex);
+            return;
+        }
+        if (getModel().getSelectedHeaderCells().size() > 0) {
+            super.doAfterSelectionChange(uiRowIndex, uiColumnIndex);
+            return;
+        }
```
Comment (karreiro) [inline kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/expressions/types/context/ContextGrid.java]: Personally, I would rather a long condition line than the having the `super.doAfterSelectionChange(..)` twice:
```java
if (getModel().getSelectedCells().stream().map(GridData.SelectedCell::getRowIndex).distinct().count() > 1 || getModel().getSelectedHeaderCells().size() > 0) {
    super.doAfterSelectionChange(uiRowIndex, uiColumnIndex);
    return;
}
```
..or maybe, we could extract these conditions:
```java
if (hasSelectedCells() || hasSelectedHeader()) {
    super.doAfterSelectionChange(uiRowIndex, uiColumnIndex);
    return;
}
```
But, feel free to keep it as is.

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/expressions/types/dtable/DecisionTableGrid.java [hunk 8]
```diff
@@ -544,6 +544,15 @@ public void setDecisionTableOrientation(final DecisionTableOrientation orientati
...
                                           final int uiColumnIndex) {
+        if (getModel().getSelectedCells().size() > 1) {
+            super.doAfterSelectionChange(uiRowIndex, uiColumnIndex);
+            return;
+        }
+        if (getModel().getSelectedHeaderCells().size() > 0) {
+            super.doAfterSelectionChange(uiRowIndex, uiColumnIndex);
+            return;
+        }
+
```
Comment (karreiro) [inline kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/expressions/types/dtable/DecisionTableGrid.java]: ..the same comment above.

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/editors/expressions/types/context/ContextGridTest.java (hunk 1 above)
Comment (manstis) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/editors/expressions/types/context/ContextGridTest.java]: @karreiro I do have similar tests for checking when `ctrl` is not pressed already :-)

These _new_ tests select one row/cell check the events that were fired and then select an additional row/cell to check the events that fire again... I guess I could remove the first assertion, but I'd still need to reset the mock as selecting the first row/cell fires the event - that I need to also verify when multiple rows/cells are then selected.

Comment (karreiro) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/editors/expressions/types/context/ContextGridTest.java]: @manstis Ohh.. I understand now. Thank you for the clarification :-)

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/expressions/types/context/ContextGrid.java (hunk 7 above)
Comment (manstis) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/expressions/types/context/ContextGrid.java]: Updated.

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/expressions/types/dtable/DecisionTableGrid.java (hunk 8 above)
Comment (manstis) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/expressions/types/dtable/DecisionTableGrid.java]: Updated.

**Comment by manstis** [comment]
@karreiro New commit with your requests added.

**Comment by manstis** [comment]
@jomarko This is ready for your re-review. Thanks.

**Pull request merged**